tortoise/verifying_tortoise.go

```diff
@@ -809,27 +809,164 @@ func (t *turtle) handleLayerBlocks(ctx *tcontext, layerID types.LayerID) error {
+		// The former is temporary and will go away after `zdist' layers. And it should be true of an entire layer,
+		// not just of a single block. The latter could cause the global opinion of a single block to permanently
+		// be abstain. As long as the contextual validity of any block in a layer is unresolved, we cannot verify
+		// the layer (since the effectiveness of each transaction in the layer depends upon the contents of the
+		// entire layer and transaction ordering). Therefore we have to enter self healing in this case.
+		// TODO: abstain only for entire layer at a time, not for individual blocks (optimization)
+		//   see https://github.com/spacemeshos/go-spacemesh/issues/2674
+		if !globalOpinionDecided {
+			logger.With().Warning("global opinion on block is abstain, cannot verify layer",
+				bid,
+				log.FieldNamed("global_opinion", globalOpinionOnBlock),
+				log.FieldNamed("local_opinion", localOpinionOnBlock))
+			return nil, nil
+		}
+	}
+	return contextualValidity, nil
+}
+
+func (t *turtle) healingTortoise(ctx *tcontext, logger log.Log, lid types.LayerID) (bool, error) {
+	// Verifying tortoise will wait `zdist' layers for consensus, then an additional `ConfidenceParam'
+	// layers until all other nodes achieve consensus. If it's still stuck after this point, i.e., if the gap
+	// between this unverified candidate layer and the latest layer is greater than this distance, then we trigger
+	// self healing. But there's no point in trying to heal a layer that's not at least Hdist layers old since
+	// we only consider the local opinion for recent layers.
+	logger.With().Debug("considering attempting to heal layer",
+		log.FieldNamed("layer_cutoff", t.layerCutoff()),
+		log.Uint32("zdist", t.Zdist),
+		log.FieldNamed("last_layer_received", t.Last),
+		log.Uint32("confidence_param", t.ConfidenceParam))
+	if !(lid.Before(t.layerCutoff()) && t.Last.Difference(lid) > t.Zdist+t.ConfidenceParam) {
+		return false, nil
+	}
+	logger.With().Info("start self-healing with verified layer", t.Verified)
+	lastLayer := t.Last
+	// don't attempt to heal layers newer than Hdist
+	if lastLayer.After(t.layerCutoff()) {
+		lastLayer = t.layerCutoff()
+	}
+	lastVerified := t.Verified
+	t.heal(ctx, lastLayer)
+	logger.With().Info("finished self-healing with verified layer", t.Verified)
+
+	if !t.Verified.After(lastVerified) {
+		// if self healing didn't make any progress, there's no point in continuing to attempt
+		// verification
+		// give up trying to verify layers and keep waiting
+		logger.With().Info("failed to verify candidate layer, will reattempt later")
+		return false, nil
+	}
+
+	// reinitialize context because contextually valid blocks were changed
+	ctx = wrapContext(ctx)
+	// rescore goodness of blocks in all intervening layers on the basis of new information
+	for layerID := lastVerified.Add(1); !layerID.After(t.Last); layerID = layerID.Add(1) {
+		if err := t.scoreBlocksByLayerID(ctx, layerID); err != nil {
```

Comment:
no, we rescore after healing because local opinion changed and goodness may have changed as well. doing this before will not change goodness. 

as for loading blocks i think it was already discussed, we can't do it just any time when healing is called so we will need to have external signal (state sync) or just do it periodically.